extra/grpc/sam/sam.py

```diff
@@ -0,0 +1,212 @@
+#! /usr/bin/env python3
+from __future__ import annotations
+
+from concurrent import futures
+import argparse
+import os
+import signal
+import sys
+import os
+import time
+
+import backend_pb2
+import backend_pb2_grpc
+
+import grpc
+
+import torch
+from functools import partial
+from segment_anything_hq import SamAutomaticMaskGenerator
+from segment_anything_hq.modeling import ImageEncoderViT, MaskDecoderHQ, PromptEncoder, Sam, TwoWayTransformer
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+_ONE_DAY_IN_SECONDS = 60 * 60 * 24
+PROMT_EMBED_DIM=256
+IMAGE_SIZE = 1024
+VIT_PATCH_SIZE=16
+
+# Enum for sam model type
+class SamModelType:
+    default = "sam_hq_vit_h.pth"
+    vit_h = "sam_hq_vit_h.pth"
+    vit_l = "sam_hq_vit_l.pth"
+    vit_b = "sam_hq_vit_b.pth"
+    vit_tiny = "sam_hq_vit_tiny.pth"
+
+
+# If MAX_WORKERS are specified in the environment use it, otherwise default to 1
+MAX_WORKERS = int(os.environ.get('PYTHON_GRPC_MAX_WORKERS', '1'))
+
+
+# Implement the BackendServicer class with the service methods
+class BackendServicer(backend_pb2_grpc.BackendServicer):
+    """
+    A gRPC servicer for the backend service.
+    """
+
+    def Health(self, request, context):
+        return backend_pb2.Reply(message=bytes("OK", "utf-8"))
+
+    def LoadModel(self, request, context):
+        try:
+            model_name = request.model_name
+            if model_name not in SamModelType.__dict__.keys():
+                raise Exception(f"Model name {model_name} not found in {SamModelType.__dict__.keys()}")
+            model_path = request.model_path
+            # check the model_path is valid
+            if not os.path.exists(model_path):
+                raise Exception(f"Model path {model_path} does not exist")
+
+            match model_name:
+                case SamModelType.default:
+                    sam = _build_sam_vit_h(checkpoint=model_path)
+                case SamModelType.vit_h:
+                    sam = _build_sam_vit_h(checkpoint=model_path)
+                case SamModelType.vit_l:
+                    sam = _build_sam_vit_l(checkpoint=model_path)
+                case SamModelType.vit_b:
+                    sam = _build_sam_vit_b(checkpoint=model_path)
+                case SamModelType.vit_tiny:
+                    # TODO: Implement this
+                    pass
+                case _:
+                    raise Exception(f"Model name {model_name} not found in {SamModelType.__dict__.keys()}")
+            # TODO No sure if this is the right way to do it
+            self.model=sam
```

Comment:
Is it ok to save it to the model. By the way, I do not see model attributes but I see `self.model` was used at some backend.

Reply:
Never mind. I have seen it in the protobuf file.